runtimes/ruby-3.1/server.rb

```diff
@@ -3,117 +3,271 @@
+      header = header[5..-1]
+      header = header.gsub("_", "-")
+      header = header.downcase
+
+      if(!header.start_with?('x-open-runtimes-'))
+        headers[header] = value
+      end
+    end
   end
 
-  unless defined?(main = ())
-    status 500
-    return { stderr: 'File does not specify a main() function.' }.to_json
+  contentType = request.env['CONTENT_TYPE']
+  if(contentType == nil)
+    contentType = 'text/plain'
   end
 
+  if(contentType.include?('application/json'))
+    if(!(bodyString.empty?))
+      body = JSON.parse(bodyString)
+    end
+  end
+
+  contextReq = Request.new(bodyString, body, headers, method, url, path, port, scheme, host, query, queryString)
+  contextRes = Response.new
+  context = Context.new(contextReq, contextRes)
+
+  customstd = nil
+  
+  output = nil
+
   begin
+    load(USER_CODE_PATH + '/' + ENV['OPEN_RUNTIMES_ENTRYPOINT'])
+
+    unless defined?(main = ())
+      raise 'User function is not valid.'
+    end
+
     system_out = $stdout
     system_err = $stderr
-    user_out = StringIO.new
-    user_err = StringIO.new
-    $stdout = user_out
-    $stderr = user_err
+    customstd = StringIO.new
+    $stdout = customstd
+    $stderr = customstd
 
-    user_response = main(requestData, runtimeResponse)
-
-    response = {
-      response: user_response,
-      stdout: user_out.string
-    }.to_json
+    output = main(context)
   rescue Exception => e
-    status 500
-    return {
-      stdout: user_out.string,
-      stderr: user_err.string + "\n" + e.backtrace.join("\n")
-    }.to_json
+    context.error(e)
+    context.error(e.backtrace.join("\n"))
+    output = context.res.send('', 500, {})
   ensure
     $stdout = system_out
     $stderr = system_err
-    user_out = nil
-    user_err = nil
   end
 
-  status 200
+  if output == nil
+    context.error('Return statement missing. return context.res.empty() if no response is expected.')
+    output = context.res.send('', 500, {})
+  end
+
+  if(output['body'] == nil)
+    output['body'] = ''
+  end
+
+  if(output['statusCode'] == nil)
+    output['statusCode'] = 200
+  end
+
+  if(output['headers'] == nil)
+    output['headers'] = {}
+  end
+
+  output['headers'].each do |header, value|
+    if !header.downcase.start_with?('x-open-runtimes-')
+      response.headers[header.downcase] = value
+    end
+  end
+
+  if customstd.string != nil && !customstd.string.empty?
+    context.log('Unsupported log noticed. Use context.log() or context.error() for logging.')
+  end
+
+  response.headers['x-open-runtimes-logs'] = ERB::Util.url_encode(context.logs.join('\n'))
+  response.headers['x-open-runtimes-errors'] = ERB::Util.url_encode(context.errors.join('\n'))
+
+  if output['headers']['content-type'] != nil
+    response.content_type = output['headers']['content-type']
+  end
+
+  response.status = output['statusCode']
+  response.body = output['body']
+end
+
+get '*' do
+  handle(request, response)
+  return response
+end
```

Comment:
If you return the response from handle, then you can omit the separate return for each route:

```suggestion
  response.body = output['body']
  response
end

get '*' do
  return handle(request, response)
end
```

Reply:
Applied